azavea/iow-boundary-tool
# Add Support for Uploading Files

## Overview

Adds back-end support for uploading and saving files.

~~All uploaded files are saved to S3. It was easier to do this than to have development save locally. The development bucket will be shared between all developers and test environments.~~

Uploaded files are saved to local disk in development, and in S3 buckets in staging and production.

Supports multiple reference image uploads, and a single shape file upload. Supports uploading a zipped shapefile, as well as a geojson file.

Reference images are reported as signed S3 URLs, so the front-end can access them despite the buckets themselves being private. These URLs are valid for about an hour.

The front-end is updated to upload the files correctly, and also to load the shapes and reference images from the back-end. An edge case with L.DistortableImage was handled, without which navigating away from the map was not possible.

Also refactors the tests, updates the reference image tests to upload files, and adds tests for shapefile and geojson uploads.

~~More UI polish will come with #158.~~

Closes #81 
Closes #158 

### Demo

https://user-images.githubusercontent.com/1430060/199588222-5961e7e9-6963-4288-9b6e-26c1c1f88d4b.mp4

## Testing Instructions

- Check out this PR, run `update` and `resetdb` and `server`
- Log in as a contributor, go through the Welcome modals
- Select a number of files for uploading
  - ℹ️ Use the shapefile or geojson added for testing if you don't have any locally
  - ℹ️ Use any images available for testing reference images
  - [x] Ensure that the uploads are handled correctly
  - [x] Ensure that the reference images are rendered correctly
  - [ ] Ensure that the shapefile / geojson is ingested correctly
- Navigate around the app
  - [ ] Make sure there's no other crashes

## Checklist
